### ISSUE TITLE
ERROR: Unexpected error in munkilib.installer when installing 3rd Party and Apple Updates in same run

### ISSUE THREAD

**Issue by miawri** [munki/munki#998]
Installing macOS updates that require a reboot alongside 3rd Party installs (MS Office) are possibly causing a race condition which results in the 3rd party install failing.

I've attached a relevant ManagedSoftwareUpdate.log and install.log that appears to show the OS starting the Office install and then rebooting due to the (in this case) Security Update.
I am currently working around the issue by requiring the respective OS be on the latest build number before instlling the Office update itself.

[installLogOutput.log](https://github.com/munki/munki/files/4814153/installLogOutput.log)
[mscLogOutput.log](https://github.com/munki/munki/files/4814156/mscLogOutput.log)

**Comment by gregneagle** [comment]
There's some weird time jumps in the install.log that makes following the events difficult, but I think you are correct about what is happening.

This might be difficult to prevent 100% because there's really no way for Munki to know that the user has triggered a "Software Update" update.

I think the two things we can do are:

1) When a user clicks the MSC button that opens the Software Update preferences pane, write a flag/trigger file that causes Munki to skip running at the loginwindow (at least the next time we're at the loginwindow). We then have to be careful about cleaning that up if for whatever reason that's present if the user approves a "normal" Munki install that requires a logout or restart.

2) When running at the loginwindow, attempt to detect if a Software Update process is running and exit if so.

This points out why having one process (managedsoftwareupdate) handle all the updates is better than having two that could interfere with each other.  Apple does not make our jobs easier. :-(

**Comment by gregneagle** [comment]
Looks like `suhelperd` and `softwareupdated` might be running all the time:

```
$ ps auxwww | grep "Software Update.app/Contents/Resources/"
root               615   0.0  0.0  4348556    624   ??  Ss    2Jun20   0:03.67 /System/Library/CoreServices/Software Update.app/Contents/Resources/suhelperd
_softwareupdate    612   0.0  0.0  4495136   3024   ??  Ss    2Jun20  14:16.29 /System/Library/CoreServices/Software Update.app/Contents/Resources/softwareupdated
```

**Comment by gregneagle** [comment]
So maybe we need to look for `/System/Library/CoreServices/Software Update.app/Contents/Resources/Software Update` and `/usr/sbin/softwareupdate`

**Comment by gregneagle** [comment]
We should probably look before beginning any installations and exit if either process is running. This might cause some (strictly unnecessary) aborted installations, but I'm not sure there's a safe alternative.

**Comment by miawri** [comment]
Apple makes it hard indeed...
I agree that the best workaround would be to check whether either of the processes are running and bail if so.  I don't really see a safer/better/too-overly-complicated alternative.

**Comment by poundbangbash** [comment]
Checking for the running processes may not be enough. There could be a timing issue if the user is directed to the System Preference software update but doesn't engage that update immediately.  How long should munki wait to see if a Software Update is going to process the 3rd party install?  If the 3rd party install happens immediately because the Software Update process hasn't started but then the user clicks install on the Software Update a short time later, the SU may finish and restart before the 3rd party installs complete and the same issue happens just further along the 3rd party install timeline.

I'm not sure what the right angle of attack is for this, but wanted to mention the above.

-Eric

**Comment by gregneagle** [comment]
I'm not proposing waiting for anything at all. I'm proposing that when managedsoftwareupdate runs, that it checks (perhaps at initial launch, and then again before starting any pending installs, or possibly just the latter) to see if either of those processes are running, and if so, exiting (or _skipping_ the install phase).

This is likely to cause all sorts of annoying behaviors if the admin has configured Software Update to run automatically, as I can imagine a scenario where a user uses an optional_install item to install some app, but the install is aborted because Apple's software update process just happened to fire while managedsoftwareupdate was running.

Maybe this change should happen only when Munki is running at the loginwindow.

**Comment by gregneagle** [comment]
As of this commit: https://github.com/munki/munki/commit/f5ff3d86343c597904217a30225ce296c78066e0
Managed Software Center is doing some things to make this scenario less likely: specifically, it's attempting to make sure that the various trigger files that cause `managedsoftwareupdate` to run at the loginwindow have been removed.
Of course there are reasons this might not be 100% sufficient in preventing this problem. I need to figure out how to reliably determine if softwareupdate/Software Update is running while we're at the loginwindow, and so far I've been unsuccessful.

**Comment by gregneagle** [comment]
Partially addressed in 5.0.1 release.
